core/src/Streamly/Internal/Data/Stream/Exception.hs

```diff
@@ -350,17 +354,81 @@ bracketIO3 bef aft onExc onGC =
         onGC
         (inline MC.try)
 
--- | Run a monadic effect supplying it with a function to register cleanup
--- actions that are automatically invoked on exception or after the effect
--- function is done.
-{-# INLINE cleanupEffectIO #-}
-cleanupEffectIO :: (MonadIO m, MonadCatch m) =>
-    ((IO () -> IO ()) -> m ()) -> m ()
-cleanupEffectIO action = do
+-- | @withBracketIOM action@ runs the given @action@, providing it with a
+-- special function called @registerHook@ as argument. A @registerHook alloc
+-- free@ call can be used within @action@ any number of times to allocate
+-- resources that are freed automatically when 'withBracketIOM' ends or if an
+-- exception occurs at any time. @alloc@ is a function used to allocate a
+-- resource and @free@ is to free the allocated resource. @registerHook@
+-- returns the allocated resource and a @release@ function, it ensures safe
+-- allocation of resource and sets up its automatic release on exception or at
+-- the end. The @release@ function returned by @registerHook@ can be used to
+-- free the resource manually at any time.
+--
+-- This provides similar functionality as the bracket function in the base
+-- library but it is more powerful as resources can be allocated at any time
+-- within the scope and can be released at any time.
+--
+-- This function does not rely on GC for cleanup. Cleanup is deterministic.
+--
+-- Exception safe, thread safe.
+{-# INLINE withBracketIOM #-}
+withBracketIOM :: (MonadIO m, MonadCatch m) =>
+    ((IO b -> (b -> IO c) -> IO (b, IO ())) -> m a) -> m a
+withBracketIOM action = do
+    ref <- liftIO $ newIORef (0 :: Int, Map.empty)
+    -- XXX use mask or MC.finally?
+    r <- action (bracket ref) `MC.onException` aft ref
+    aft ref
+    return r
+
+    where
+
+    aft ref = liftIO $ do
+        xs <- readIORef ref
+        sequence_ (snd xs)
+
+    bracket ref alloc free = do
+        (r, index) <- liftIO $ mask_ $ do
+            r <- alloc
+            idx <- atomicModifyIORef ref (\(i, mp) ->
+                ((i + 1, Map.insert i (void $ free r) mp), i))
+            return (r, idx)
+
+        let free1 =
+                atomicModifyIORef ref (\(i, mp) ->
+                    ((i, Map.delete index mp), ()))
+        return (r, free1)
```

Comment:
This only deletes the registered free function at the index.
Shouldn't it also run the `free` function?

Reply:
Also, I don't see how we can call the free function manually. Can you give an example or use it in the tests?